<a href="https://colab.research.google.com/github/morreh89/Moringa_IP/blob/master/Moringa_Data_Science_Prep_W4_Independent_Project_2020_07_Maurice_Kimanzi_Python_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **Python IP Week 4: Processing the dataset for an electric car-sharing service company to understand electric car usage over time** **by solving for the following research question:**


***1.1 Importing python Libraries***

In [1]:
# Importing the pandas library

import pandas as pd

# Importing the numpy library

import numpy as np

# importing the datetime library

from datetime import datetime

***1.2 Reading the Dataset from our CSV file url***

In [2]:
#load the dataset from an external url
url =('http://bit.ly/autolib_dataset')
df = pd.read_csv(url)
#previewing the dataset
df.head(5)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23

[5 rows x 25 columns]

***1.3 Let's access information about the dataset now***

---




In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   Displayed comment    111 non-null    object
 9   ID                   5000 non-null   object
 10  Kind                 5000 non-null   object
 11  Geo point            5000 non-null   object
 12  Postal code          5000 non-null   int64 
 13  Public name          5000 non-null   object
 14  Rental status        5000 non-null   object
 15  Scheduled at         47 non-null     object
 16  Slots 

In [4]:
df.shape

(5000, 25)

In [5]:
df.columns

Index(['Address', 'Cars', 'Bluecar counter', 'Utilib counter',
       'Utilib 1.4 counter', 'Charge Slots', 'Charging Status', 'City',
       'Displayed comment', 'ID', 'Kind', 'Geo point', 'Postal code',
       'Public name', 'Rental status', 'Scheduled at', 'Slots', 'Station type',
       'Status', 'Subscription status', 'year', 'month', 'day', 'hour',
       'minute'],
      dtype='object')

***1.4 Let's clean the dataset***


    1.4.1 Improving validity of the dataset

In [6]:
#Dropping columns that we won't need for this research

df.drop(['Displayed comment', 'Scheduled at', 'ID', 'Geo point','Address', 'Cars'], axis = 1, inplace = True)
df.head(5)

Bluecar counter  Utilib counter  Utilib 1.4 counter  ...  day hour minute
0                0               0                   0  ...    8   11     43
1                6               0                   0  ...    6    7     24
2                3               0                   2  ...    3   20     14
3                3               1                   0  ...    4    4     37
4                3               0                   0  ...    8   17     23

[5 rows x 19 columns]

In [7]:
# Data Cleaning Action: checking for outliers using IQR Scores


Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)

IQR = Q3 - Q1


(df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))



Bluecar counter  Charge Slots  Charging Status  ...  minute  month   year
0               False         False            False  ...   False  False  False
1               False         False            False  ...   False  False  False
2               False         False            False  ...   False  False  False
3               False          True            False  ...   False  False  False
4               False         False            False  ...   False  False  False
...               ...           ...              ...  ...     ...    ...    ...
4995            False         False            False  ...   False  False  False
4996            False          True            False  ...   False  False  False
4997            False          True            False  ...   False  False  False
4998            False         False            False  ...   False  False  False
4999            False         False            False  ...   False  False  False

[5000 rows x 19 columns]

In [8]:
df.shape

(5000, 19)

***1.4.2 Checking and Improving on Completeness***


In [9]:
#checking for null columns
df.isnull().sum()

Bluecar counter        0
Utilib counter         0
Utilib 1.4 counter     0
Charge Slots           0
Charging Status        0
City                   0
Kind                   0
Postal code            0
Public name            0
Rental status          0
Slots                  0
Station type           0
Status                 0
Subscription status    0
year                   0
month                  0
day                    0
hour                   0
minute                 0
dtype: int64

In [10]:
#checking for not null columns
df.notnull().sum()

Bluecar counter        5000
Utilib counter         5000
Utilib 1.4 counter     5000
Charge Slots           5000
Charging Status        5000
City                   5000
Kind                   5000
Postal code            5000
Public name            5000
Rental status          5000
Slots                  5000
Station type           5000
Status                 5000
Subscription status    5000
year                   5000
month                  5000
day                    5000
hour                   5000
minute                 5000
dtype: int64

In [11]:
#preview the dataset once again
df.head(5)

Bluecar counter  Utilib counter  Utilib 1.4 counter  ...  day hour minute
0                0               0                   0  ...    8   11     43
1                6               0                   0  ...    6    7     24
2                3               0                   2  ...    3   20     14
3                3               1                   0  ...    4    4     37
4                3               0                   0  ...    8   17     23

[5 rows x 19 columns]

***1.4.3 Improving Consistency of our dataset***


In [12]:
#check if there are duplicates
df.duplicated().sum()
#there are no duplicates in our dataset

0

In [13]:
#check again our dataset shape
df.shape

(5000, 19)

***1.4.4 Enhancing Uniformity***
  
Changing the case of column names to upper, stripping whitespaces, removing periods and replacing spaces with underscore


In [14]:
#converting all column names to uppercase to enhance uniformity and adding '_'

df.columns = df.columns.str.upper().str.strip().str.replace(' ', '_').str.replace('.', '')

df.columns

Index(['BLUECAR_COUNTER', 'UTILIB_COUNTER', 'UTILIB_14_COUNTER',
       'CHARGE_SLOTS', 'CHARGING_STATUS', 'CITY', 'KIND', 'POSTAL_CODE',
       'PUBLIC_NAME', 'RENTAL_STATUS', 'SLOTS', 'STATION_TYPE', 'STATUS',
       'SUBSCRIPTION_STATUS', 'YEAR', 'MONTH', 'DAY', 'HOUR', 'MINUTE'],
      dtype='object')

***1.5 Exporting the cleaned dataset as df1***


In [15]:
df.to_csv('df1.csv')


In [16]:
df1 = pd.read_csv('df1.csv', index_col=0, encoding='utf-8')
df1.head(5)

BLUECAR_COUNTER  UTILIB_COUNTER  UTILIB_14_COUNTER  ...  DAY HOUR MINUTE
0                0               0                  0  ...    8   11     43
1                6               0                  0  ...    6    7     24
2                3               0                  2  ...    3   20     14
3                3               1                  0  ...    4    4     37
4                3               0                  0  ...    8   17     23

[5 rows x 19 columns]

***1.6 Answering Research Questions***

> **Q1: Identify the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018. **





Add three columns for the three car counters (BLUECAR COUNTER	, UTILIB COUNTER and UTILIB 1.4 COUNTER)

+ve number will mean that a car was returned -ve number will mean that a car was picked 0 will mean that nothing no car was picked/returned

In [ ]:
#Creating column for bluecar usage
df1['BLUECAR_USE'] = df1.BLUECAR_COUNTER.diff()

#Creating column for utilib usage
df1['UTILIB_USE'] = df1.UTILIB_COUNTER.diff()

#Creating column for utilib 14 usage
df1['UTILIB_14_USE'] = df1.UTILIB_14_COUNTER.diff()

df1

In [18]:
df1.BLUECAR_USE.unique()

array([nan,  6., -3.,  0.,  3., -2.,  5., -6.,  1.,  4., -5., -1.,  2.,
       -4.,  7., -7.])

In [47]:
#from above -7 shows most Blucars picked
#now we calculate the popular hour
popHour = df1[(df1.CITY == 'Paris') & (df1.BLUECAR_USE < 0)].groupby('HOUR')['HOUR'].count().sort_values(ascending = False)
popHour.head(1)



HOUR
21    80
Name: HOUR, dtype: int64

The most popular hour of the day for picking bluecars is 2100 hours i.e 9pm

**Q2: What is the most popular hour for returning cars?**


In [66]:
dfReturn = df1[df1.BLUECAR_USE > 0].groupby('HOUR')['HOUR'].count().sort_values(ascending = False)
dfReturn.head(1)


HOUR
6    101
Name: HOUR, dtype: int64


The most popular hour for returning bluecars is 0600 hours/ 6pm

**Q3: What station is the most popular overall?**




In [ ]:
popstat = df1[(df1.KIND == 'STATION') & (df1.STATUS == 'ok')].groupby('PUBLIC_NAME')[['KIND']].count().sort_values(by = 'KIND', ascending  = False)
popstat.head(1)

KIND
PUBLIC_NAME                     
Paris/Porte de Montrouge/8    13

The most popular overal station is Paris/Porte de Montrouge/8	

**Q4: What station is most popular at the most popular picking hour?**


In [37]:
stationHour = df1[(df1.KIND == 'STATION') & (df1.HOUR == 21)  & (df1.STATUS == 'ok')].groupby('PUBLIC_NAME').count()[['KIND']].sort_values(by = 'KIND', ascending = False)
stationHour.head(1)

KIND
PUBLIC_NAME            
Paris/Tronchet/19     4

The most popular station at the most popular picking hour is Paris/Tronchet/19	

**Q5: What postal code is the most popular for picking up Blue cars?**

In [ ]:
postCode = df1[df1.BLUECAR_USE < 0].groupby('POSTAL_CODE').count()[['BLUECAR_USE']].sort_values(by = 'BLUECAR_USE', ascending = False)
postCode.head(1)

BLUECAR_USE
POSTAL_CODE             
75015                116

**Q6: What postal code is the most popular for picking up Bluecars at the most popular picking hour?**


In [38]:
code = df1[(df1.BLUECAR_USE < 0) & (df1.HOUR == 21)].groupby('POSTAL_CODE').count()[['BLUECAR_USE']].sort_values(by = 'BLUECAR_USE', ascending = False)
code.head(1)

BLUECAR_USE
POSTAL_CODE             
75008                  9

**Q7: Does the most popular station belong to that postal code overall?**
    No.

In [41]:
df1[df1.PUBLIC_NAME == 'Paris/Porte de Montrouge/8'][['PUBLIC_NAME', 'POSTAL_CODE']].head(1)

PUBLIC_NAME  POSTAL_CODE
122  Paris/Porte de Montrouge/8        75014

**Q8: Does the most popular station belong to that postal code at the most popular picking hour?**


In [42]:
df1[(df1.PUBLIC_NAME == 'Paris/Tronchet/19') & (df1.HOUR == 21)][['PUBLIC_NAME', 'POSTAL_CODE']].head(1)

PUBLIC_NAME  POSTAL_CODE
2172  Paris/Tronchet/19        75008

The most popular station does not belong to that postal code. It belongs to postal code 75008



## **Do the results change if you consider Utilib and Utilib 1.4 instead of Blue cars?**

**Using Utilib 1.4**



**Q1:** **What is the most popular hour of the day for picking up a shared electric car in the city of Paris over the month of April 2018 ? (working on Utilib 1.4)**

In [44]:
df1[(df1.CITY == 'Paris') & (df1.UTILIB_14_USE < 0 )].groupby('HOUR')['HOUR'].count().sort_values(ascending = False).head(1)

HOUR
3    20
Name: HOUR, dtype: int64

**Q2: What is the most popular hour for returning cars?**


In [50]:
 df1[(df1.CITY == 'Paris') & (df1.UTILIB_14_USE > 0)].groupby('HOUR')['HOUR'].count().sort_values(ascending = False).head(1)

HOUR
9    17
Name: HOUR, dtype: int64


**Q3: What station is the most popular overall ?**

In [51]:
df1[(df1.KIND == 'STATION') & (df1.STATUS == 'ok')].groupby('PUBLIC_NAME')[['KIND']].count().sort_values(by = 'KIND', ascending  = False).head(1)


KIND
PUBLIC_NAME                     
Paris/Porte de Montrouge/8    13

**Q4: What is the most popular station at the most popular hour ?**


In [53]:
df1[(df1.KIND == 'STATION') & (df1.HOUR == 3)  & (df1.STATUS == 'ok')].groupby('PUBLIC_NAME').count()[['KIND']].sort_values(by = 'KIND', ascending = False).head(1)


KIND
PUBLIC_NAME           
Paris/Mortier/15     2

**Q5: What postal code is most popular when picking up Utilib cars?**


In [54]:
df1[df1.BLUECAR_USE < 0].groupby('POSTAL_CODE').count()[['UTILIB_14_USE']].sort_values(by = 'UTILIB_14_USE', ascending = False).head(1)


UTILIB_14_USE
POSTAL_CODE               
75015                  116

**Q6: What postal code is the most popular for picking up Utilib cars at the most popular picking hour?**


In [55]:
df1[(df1.UTILIB_14_USE < 0) & (df1.HOUR == 21)].groupby('POSTAL_CODE').count()[['UTILIB_14_USE']].sort_values(by = 'UTILIB_14_USE', ascending = False).head(1)


UTILIB_14_USE
POSTAL_CODE               
75017                    3

**Q7: Does the most popular station belong to that postal code overall?**


In [56]:
df1[df1.PUBLIC_NAME == 'Paris/Porte de Montrouge/8'][['PUBLIC_NAME', 'POSTAL_CODE']].head(1)

PUBLIC_NAME  POSTAL_CODE
122  Paris/Porte de Montrouge/8        75014

**Q8: Does the most popular station belong to that postal code at the most popular picking hour?**


In [61]:
df1[(df1.PUBLIC_NAME == 'Paris/Mortier/15') & (df1.HOUR == 3)][['PUBLIC_NAME', 'POSTAL_CODE']].head(1)

PUBLIC_NAME  POSTAL_CODE
139  Paris/Mortier/15        75020